### ISSUE TITLE
The Kblam performs poorly in refusal mode.

### ISSUE THREAD

**Issue by yujia-fu** [microsoft/KBLaM#67]
I trained a model using a synthetic dataset with the hyperparameters specified in the paper, and evaluated it on the Enron dataset. In refusal mode, KBlam failed to demonstrate the ability to refuse answers, showing a  difference in performance compared to ICL. Even when tested on synthetic data, KB still fails to demonstrate the ability to refuse answers.

kb: {'precision': 0.0, 'recall': 0.0, 'num_samples': 100}
icl: {'precision': 0.8260869565217391, 'recall': 0.95, 'num_samples': 100}

Some hyperparameters were set as follows:
python eval.py refusal \
--dataset_dir enron \
 --llm_base_dir Meta-Llama-3-8B-Instruct \
--llm_type llama3 \
--save_dir eval_llama \
--test_dataset enron \
--kb_size 100 \
--seed 42 \
--eval_mode kb \
--kb_layer_frequency 3

**Comment by yujia-fu** [comment]
The data sampling logic needs to be modified if using a single machine with multiple cards for training. Otherwise the uneven distribution of data types per gpu will result in a model saved by the main process that has only a generation effect but no rejection effect.

**Comment by ltp-THU** [comment]
I encounter the same low refusal performance with phi3 model. The KB does not refuse at all. My question is to train the model for refusal task, the training data should be modified? The paper seems train one model to process multiple tasks. The model is supposed to be "over-refusal" as displayed in the paper.
